File: app/src/main/java/com/example/breezil/favnews/ui/PrefFragment.java
package com.example.breezil.favnews.ui;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceFragment;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v14.preference.MultiSelectListPreference;
import android.support.v4.view.ViewCompat;
import android.support.v7.preference.ListPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceGroup;
import android.support.v7.preference.PreferenceManager;
import android.view.View;
import android.widget.ListView;

import com.example.breezil.favnews.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public class PrefFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private MultiSelectListPreference mSourcePref;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences);

        PreferenceManager.setDefaultValues(Objects.requireNonNull(getActivity()), R.xml.preferences, false);

        PreferenceGroup preferenceGroup = (PreferenceGroup) getPreferenceScreen().getPreference(0);

        mSourcePref = (MultiSelectListPreference) preferenceGroup.getPreference(0);

        initSummary(getPreferenceScreen());
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final ListView listView = view.findViewById(android.R.id.list);
        if (listView != null)
            ViewCompat.setNestedScrollingEnabled(listView, true);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (!key.equals(getString(R.string.pref_source_key))) {
            updateSummary(findPreference(key));

        } else {
            updateMultiSummary(findPreference(key),
                    sharedPreferences.getStringSet(getString(R.string.pref_source_key), null));
        }
    }

    private void initSummary(Preference p) {
        if (p instanceof PreferenceGroup) {
            PreferenceGroup pGrp = (PreferenceGroup) p;
            for (int i = 0; i < pGrp.getPreferenceCount(); i++) {
                initSummary(pGrp.getPreference(i));
            }
        } else {
            updateSummary(p);
            updateMultiSummary(p, PreferenceManager.getDefaultSharedPreferences(Objects.requireNonNull(getActivity()))
                    .getStringSet(getString(R.string.pref_source_key), null));

        }
    }

    private void updateSummary(Preference p) {
        if (p instanceof ListPreference) {
            ListPreference listPref = (ListPreference) p;
            p.setSummary(listPref.getEntry());

        }
    }

    private void updateMultiSummary(Preference p, Set<String> value) {
        if (p instanceof MultiSelectListPreference) {
            MultiSelectListPreference multiSelectListPreference = (MultiSelectListPreference) p;

            List<String> entries = new ArrayList<>(value);
            StringBuilder allEntries = new StringBuilder();

            for (int i = 0; i < entries.size(); i++) {
                allEntries.append(multiSelectListPreference.getEntries()[multiSelectListPreference.findIndexOfValue(entries.get(i))])
                        .append(", ");
            }

            if (allEntries.length() > 0) {
                allEntries.deleteCharAt(allEntries.length() - 2);
            }

            p.setSummary(allEntries.toString());

        }
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        super.onPause();
    }

    @Override
    public void onStop() {
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        super.onStop();
    }
}

File: app/src/main/java/com/fave/breezil/fave/ui/preference/PrefFragment.kt
package com.fave.breezil.fave.ui.preference

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v14.preference.MultiSelectListPreference
import android.support.v14.preference.SwitchPreference
import android.support.v4.app.FragmentActivity
import android.support.v4.view.ViewCompat
import android.support.v7.preference.ListPreference
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceFragmentCompat
import android.support.v7.preference.PreferenceGroup
import android.support.v7.preference.PreferenceManager
import android.view.View
import android.widget.ListView

import com.fave.breezil.fave.R

import java.util.ArrayList
import java.util.Objects

class PrefFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    private var mSourcePref: MultiSelectListPreference? = null


//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        addPreferencesFromResource(R.xml.preferences)
//
//
//
//        PreferenceManager.setDefaultValues(Objects.requireNonNull<FragmentActivity>(activity), R.xml.preferences, false)
//
//        val preferenceGroup = preferenceScreen.getPreference(0) as PreferenceGroup
//
//        mSourcePref = preferenceGroup.getPreference(0) as MultiSelectListPreference
//
//        initSummary(preferenceScreen)
//    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences,rootKey)
        PreferenceManager.setDefaultValues(Objects.requireNonNull<FragmentActivity>(activity), R.xml.preferences, false)

        val preferenceGroup = preferenceScreen.getPreference(0) as PreferenceGroup

        mSourcePref = preferenceGroup.getPreference(0) as MultiSelectListPreference

        initSummary(preferenceScreen)

    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val listView = view.findViewById<ListView>(android.R.id.list)
        if (listView != null)
            ViewCompat.setNestedScrollingEnabled(listView, true)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (key != getString(R.string.pref_source_key)) {
            updateSummary(findPreference(key))
            updateNightMode(findPreference(key))
            restartActivity()
        } else {
            updateMultiSummary(findPreference(key),
                    sharedPreferences.getStringSet(getString(R.string.pref_source_key), null))
        }
    }

    private fun initSummary(p: Preference) {
        if (p is PreferenceGroup) {
            for (i in 0 until p.preferenceCount) {
                initSummary(p.getPreference(i))
            }
        } else {
            updateSummary(p)
            updateMultiSummary(p, PreferenceManager.getDefaultSharedPreferences(Objects.requireNonNull<FragmentActivity>(activity))
                    .getStringSet(getString(R.string.pref_source_key), null))
            updateNightMode(p)
        }
    }

    private fun updateSummary(p: Preference) {
        if (p is ListPreference) {
            p.setSummary(p.entry)

        }
    }

    private fun updateMultiSummary(p: Preference, value: Set<String>?) {
        if (p is MultiSelectListPreference) {

            val entries = ArrayList(value!!)
            val allEntries = StringBuilder()

            for (i in entries.indices) {
                allEntries.append(p.entries[p.findIndexOfValue(entries[i])])
                        .append(", ")
            }

            if (allEntries.isNotEmpty()) {
                allEntries.deleteCharAt(allEntries.length - 2)
            }

            p.setSummary(allEntries.toString())

        }
    }

    private fun updateNightMode(p: Preference) {
        if (p is SwitchPreference) {
            if (p.isChecked) {
                p.setSummary(getString(R.string.pref_theme_checked_summary))
                p.setDefaultValue(getString(R.string.pref_theme_true_value))
            } else {
                p.setSummary(getString(R.string.pref_theme_unchecked_summary))
                p.setDefaultValue(getString(R.string.pref_theme_false_value))
            }
        }
    }

    override fun onResume() {
        super.onResume()
        preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        super.onPause()
    }

    override fun onStop() {
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        super.onStop()
    }

    private fun restartActivity() {
        val intent = Intent(activity, SettingsActivity::class.java)
        startActivity(intent)
        activity!!.finish()
    }
}
